### PR TITLE
FatDB

### PR DESCRIPTION
**Pull request by debris**
First attempt to implement FatDB. I've tried to approach the problem in few different ways and this architecture seems to be the simplest (It's actually very close to c++ implementation).

changes:
- added `FatDB` structure implementing `Trie`
- added `FatDBMut` structure implementing `TrieMut`
- added `iter(&self)` method to `Trie` triat
- separated `TrieMut` from `Trie`
- added `insert_aux(&mut self, _hash: Vec<u8>, _value: Vec<u8>)` and `get_aux(&self, _hash: &[u8]) -> Option<Vec<u8>>` and their default implementations to `HashDB` triat.
- added `TrieFactory`

In next pr I will integrate it with `ethcore` and cli.

### CODE REVIEW

File: util/src/trie/fatdb.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+use hash::H256;
+use sha3::Hashable;
+use hashdb::HashDB;
+use super::{TrieDB, Trie, TrieDBIterator, TrieError};
+
+/// A mutable `Trie` implementation which hashes keys and uses a generic `HashDB` backing database.
+///
+/// Use it as a `Trie` or `TrieMut` trait object. You can use `raw()` to get the backing `TrieDB` object.
+pub struct FatDB<'db> {
```
Comment (rphmeier) [inline util/src/trie/fatdb.rs]: Comment seems wrong here.

could also add an explanation of what the fat DB does vs a normal trie

File: util/src/journaldb/archivedb.rs [hunk 2]
```diff
@@ -26,6 +26,8 @@ use kvdb::{Database, DBTransaction, DatabaseConfig};
 #[cfg(test)]
 use std::env;
 
+const AUX_FLAG: u8 = 255;
```
Comment (arkpar) [inline util/src/journaldb/archivedb.rs]: would be nice to use rocksdb columns for this eventually

File: util/src/memorydb.rs [hunk 3]
```diff
@@ -139,6 +141,13 @@ impl MemoryDB {
 		data
 	}
 
+	/// Return the internal map of auxiliary data, clearing the current state.
+	pub fn drain_aux(&mut self) -> HashMap<Bytes, Bytes> {
+		let mut aux = HashMap::new();
+		mem::swap(&mut self.aux, &mut aux);
```
Comment (arkpar) [inline util/src/memorydb.rs]: one liner with `mem:replace`

**Comment by arkpar** [comment]
What about pruning? there should be `aux_delete` as well

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: b4c4751f31f92457361226225843b51feebf94e7
  * Your subject line is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: b4c4751f31f92457361226225843b51feebf94e7
  * Your subject line is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: b4c4751f31f92457361226225843b51feebf94e7
  * Your subject line is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

File: util/src/journaldb/archivedb.rs (hunk 2 above)
Comment (debris) [inline util/src/journaldb/archivedb.rs]: I've added a description of this const and your comment. Also added this issue https://github.com/ethcore/parity/issues/1455

File: util/src/trie/fatdb.rs (hunk 1 above)
Comment (debris) [inline util/src/trie/fatdb.rs]: updated

**Comment by debris** [comment]
> What about pruning? there should be `aux_delete` as well

I've added `aux_remove`, but it's not used anywhere, cause it's implemented only for archivedb.

**Comment by gavofyork** [comment]
prefer to leave this `onice` until we've at least got 1.2.1out of the door, just to simplify backporting.

**Comment by debris** [comment]
sure

**Pull request merged**